healthz: Add onStop hook for healthz server

sayboras opened the pull request:
This is to avoid the below issue while agent is shutting down.
```
time=2025-05-26T06:11:37Z level=info msg="Signal received" signal=terminated
time=2025-05-26T06:11:37Z level=info msg="Stopping hive"
time=2025-05-26T06:12:37Z level=error msg="Stop hook context expired before job group was done" module=enterprise-agent.agent.infra.agent-healthz
time=2025-05-26T06:12:37Z level=info msg="Stop hook executed" function="*job.group.Stop (enterprise-agent.agent.infra.agent-healthz)" duration=59.999245249s
time="2025-05-26T06:12:37.446285566Z" level=fatal msg="failed to stop: context deadline exceeded" subsys=daemon
```

Reported-By: Yusuke Suzuki <yusuke.suzuki@isovalent.com>

joamaki commented on daemon/healthz/agenthealth.go:
```diff
@@ -59,47 +60,68 @@ func registerAgentHealthHTTPService(params agentHealthParams) error {
 
-	available := len(hosts)
-	for name, host := range hosts {
-		params.JobGroup.Add(job.OneShot(fmt.Sprintf("agent-healthz-server-%s", name), func(ctx context.Context, health cell.Health) error {
```
There's nothing wrong with using `OneShot` for long-running jobs (we've got tons of these). The issue is the `ctx` is not used in `Listen` thus we can't stop it when the job group is being stopped. Just changing `Listen(context.Background(), ...` to `Listen(ctx, ...`) should do the trick.

sayboras replied:
Good catch :+1:

sayboras committed:
healthz: Use the ctx from OneShot Job

This is to avoid the below issue while agent is shutting down.

```
time=2025-05-26T06:11:37Z level=info msg="Signal received" signal=terminated
time=2025-05-26T06:11:37Z level=info msg="Stopping hive"
time=2025-05-26T06:12:37Z level=error msg="Stop hook context expired before job group was done" module=enterprise-agent.agent.infra.agent-healthz
time=2025-05-26T06:12:37Z level=info msg="Stop hook executed" function="*job.group.Stop (enterprise-agent.agent.infra.agent-healthz)" duration=59.999245249s
time="2025-05-26T06:12:37.446285566Z" level=fatal msg="failed to stop: context deadline exceeded" subsys=daemon
```

Reported-By: Yusuke Suzuki <yusuke.suzuki@isovalent.com>
Suggested-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Tam Mach <tam.mach@cilium.io>

HadrienPatte commented on daemon/healthz/agenthealth.go:
Do we want to apply the same fix to the kube proxy health endpoint here?

https://github.com/cilium/cilium/blob/7ed8f8db61bf50021efd7f07007238dd3be55ceb/daemon/healthz/kube_proxy_healthz.go#L79
